Concept - Understanding Tokens (#40429)
dotnet/docs: docs/ai/conceptual/understanding-tokens.md

author: haywoodsloan
ms.topic: concept-article
ms.date: 04/12/2024

#customer intent: As a .NET developer, I want understand how large language models (LLMs) use tokens so I can add semantic analysis and text generation capabilities to my .NET projects.

---

# Understanding tokens

Tokens are the words or character sets that large language models (LLMs) decompose text into. During training, an LLM tokenizes the training text then analyzes the semantic relationships between these tokens, such as how commonly they're used together or if they're used in similar contexts. Once trained, an LLM uses those patterns and relationships to generate a sequence of output tokens based on the input sequence.

## Turning text into tokens

The first step when training an LLM or inputting a prompt is tokenization. The model breaks down the text into a set of strings, known as _tokens_. These tokens can be distinct words, character sets, or combinations of words and punctuation. The set of unique tokens a model has been trained on is referred to as its _vocabulary_.

For example, consider the following sentence:

> > I heard a dog bark loudly at a cat

Reviewer: ```suggestion
> I heard a dog bark loudly at a cat.
```

Author: I want to ommit the '.' to avoid discussing how punctuation is handeled in word-based tokenization (in reality there's a few different approaches to handling punctuation) to keep the focus of this example more straightfoward.

This text could be tokenized as:

- I
- heard
- a
- dog
- bark
- loudly
- at
- a
- cat

> The model will add only the first occurrence of the "a" token to it's vocabulary. All other occurrences can be represented by that first token.

Reviewer: ```suggestion
The model will add only the first occurrence of the "a" token to its vocabulary. All other occurrences can be represented by that first token.
```
I might have missed this earlier.

By having a sufficiently large set of training text, tokenization can compile a vocabulary of many thousands of tokens.

## Common tokenization methods

The specific tokenization method varies by LLM. Common tokenization methods include:

- **Word** tokenization (text is split into individual words based on a delimiter)
- **Character** tokenization (text is split into individual characters)
- **Subword** tokenization (text is split into partial words or character sets)

> For example, the GPT models, developed by OpenAI, use _Byte-Pair Encoding_ (BPE) which is a type of subword tokenization. OpenAI provides [a tool to visualize how text will be tokenized](https://platform.openai.com/tokenizer).

Reviewer: ```suggestion
For example, the GPT models, developed by OpenAI, use a type of subword tokenization that's known as _Byte-Pair Encoding_ (BPE). OpenAI provides [a tool to visualize how text will be tokenized](https://platform.openai.com/tokenizer).
```
This is another change for localization. "Which" doesn't always translate well. We try to stick with "that" -- or rewrite.

| Smaller tokens (character or subword tokenization) | - Allows the model to handle a wider range of inputs, such as unknown words, typos, or complex syntax<br>- Might allow the vocabulary size to be reduced, requiring less memory resources       | - A given text is broken into more tokens, requiring additional computational resources while processing<br>- Given a fixed token limit, the maximum size of the model's input and output is smaller |
| Larger tokens (word tokenization)                  | - A given text is broken into fewer tokens, requiring less computational resources while processing<br>- Given the same token limit, the maximum size of the model's input and output is larger | - Might cause an increased vocabulary size, requiring more memory resources<br>- Can limit the models ability to handle unknown words, typos, or complex syntax                                      |

## How LLMs use tokens

After completing tokenization, LLMs assign an ID to each unique token.

Consider the following text:

> I heard a dog bark loudly at a cat

> After using a word tokenization method, a model could assign token IDs like this:

Reviewer: ```suggestion
After the model uses a word tokenization method, it could assign token IDs as follows:
```

- I (1)
- heard (2)
- a (3)
- dog (4)
- bark (5)
- loudly (6)
- at (7)
- a (the "a" token is already assigned an ID of 3)
- cat (8)

> By assigning IDs, text can be represented as a sequence of token IDs. The previous sentence would be represented as [1, 2, 3, 4, 5, 6, 7, 3, 8]. The sentence "I heard a cat" would be represented as [1, 2, 3, 8].

Reviewer: ```suggestion
By assigning IDs, text can be represented as a sequence of token IDs. The example sentence would be represented as [1, 2, 3, 4, 5, 6, 7, 3, 8]. The sentence "I heard a cat" would be represented as [1, 2, 3, 8].
```

> As training continues, the model add any new tokens in the training text to it's vocabulary and assigns it an ID. For example:

Reviewer: ```suggestion
As training continues, the model adds any new tokens in the training text to its vocabulary and assigns it an ID. For example:
```

- meow (9)
- run (10)

> The semantic relationships between the tokens can be analyzed using these token ID sequences. Multi-valued numeric vectors, known as [_embeddings_]<!-- (understanding-embeddings.md) -->, are used to represent these relationships. An embedding is assigned to each token based on how commonly it's used together with or in similar contexts to the other tokens.

Reviewer: ```suggestion
The semantic relationships between the tokens can be analyzed by using these token ID sequences. Multi-valued numeric vectors, known as [_embeddings_]<!-- (understanding-embeddings.md) -->, are used to represent these relationships. An embedding is assigned to each token based on how commonly it's used together with, or in similar contexts to, the other tokens.
```

> Once trained, a model can calculate an embedding for a text containing multiple tokens. The model tokenizes the text, then calculates an overall embeddings value based on the learned embeddings of the individual tokens. This technique can be used for semantic document searches or adding [memories](/semantic-kernel/memories/) to an AI.

Reviewer: ```suggestion
After it's trained, a model can calculate an embedding for a text that contains multiple tokens. The model tokenizes the text, then calculates an overall embeddings value based on the learned embeddings of the individual tokens. This technique can be used for semantic document searches or adding [memories](/semantic-kernel/memories/) to an AI.
```

> During output generation, the model predicts a vector value for the next token in the sequence. The model then selects the next token from it's vocabulary based on this vector value. In practice, the models calculates multiple vectors using various elements of the preceding tokens' embeddings. The model then evaluates all potential tokens from these vectors and selects the most probable one to continue the sequence.

Reviewer: ```suggestion
During output generation, the model predicts a vector value for the next token in the sequence. The model then selects the next token from it's vocabulary based on this vector value. In practice, the model calculates multiple vectors by using various elements of the previous tokens' embeddings. The model then evaluates all potential tokens from these vectors and selects the most probable one to continue the sequence.
```
After this change, this now means “the embeddings of the previous tokens,” which is what I assume the possessive punctuation means here. If we mean, “the last action of creating token embeddings,” we could use: “the preceding token embeddings.”

Output generation is an iterative operation. The model appends the predicted token to the sequence so far and uses that as the input for the next iteration, building the final output one token at a time.

### Token limits

> LLMs have limits on the maximum number of tokens that can be used as input or generated as output. This limit often combines the input and output tokens into a maximum context window.

Reviewer: ```suggestion
LLMs have limitations regarding the maximum number of tokens that can be used as input or generated as output. This limitation often causes the input and output tokens to be combined into a maximum context window.
```
You don’t have limits, you have limitations (inherent). A regulatory agency imposes limits; you obey them. If the LLM is imposing the limits, we can use, "LLMs put limits on the..." and we can leave the "This limit often combines..." sentence as is.

For example, GPT-4 supports up to 8,192 tokens of context. The combined size of the input and output tokens can't exceed 8,192.

> Together a model's token limit and tokenization method determine the maximum length of text that can be provided as input or generated as output.

Reviewer: ```suggestion
Taken together, a model's token limit and tokenization method determine the maximum length of text that can be provided as input or generated as output.
```

> For example, consider a model with a maximum context window of 100 tokens. For the following input text:

Reviewer: ```suggestion
For example, consider a model that has a maximum context window of 100 tokens. The model processes our example sentences as input text:
```

> > I heard a dog bark loudly at a cat

Reviewer: ```suggestion
> I heard a dog bark loudly at a cat.
```

> With a word-based tokenization method, the input is 9 tokens leaving 91 **word** tokens available for the output.

Reviewer: ```suggestion
By using a word-based tokenization method, the input is nine tokens. This leaves 91 **word** tokens available for the output.
```

> With a character-based tokenization method, the input is 34 tokens (including spaces) leaving only 66 **character** tokens available for the output.

Reviewer: ```suggestion
By using a character-based tokenization method, the input is 34 tokens (including spaces). This leaves only 66 **character** tokens available for the output.
```

> ### Token based pricing and rate limiting

Reviewer: ```suggestion
### Token-based pricing and rate limiting
```

Generative AI services often use token-based pricing. The cost of each request depends on the number of input and output tokens. The pricing might differ between input and output. For example, see [Azure OpenAI Service pricing](https://azure.microsoft.com/pricing/details/cognitive-services/openai-service/).

> Generative AI services might also have limits on the maximum number of tokens per minute (TPM). These rate limits can vary depending on the service region and LLM. For more information about specific regions, see [Azure OpenAI Service quotas and limits](/azure/ai-services/openai/quotas-limits#regional-quota-limits).

Reviewer: ```suggestion
Generative AI services might also be limited regarding the maximum number of tokens per minute (TPM). These rate limits can vary depending on the service region and LLM. For more information about specific regions, see [Azure OpenAI Service quotas and limits](/azure/ai-services/openai/quotas-limits#regional-quota-limits).
```
Same change as above, depending on meaning.

## Related content

- [How Generative AI and LLMs work]<!-- (how-genai-and-llms-work.md) -->
- [Understanding embeddings]<!-- (understanding-embeddings.md) -->

> - [Working with vector databases]<!-- (working-with-vector-dbs.md) -->

Reviewer: ```suggestion
- [Working with vector databases]<!-- (working-with-vector-dbs.md) -->

[!INCLUDE [Third-party disclaimer](../../includes/third-party-contact-disclaimer.md)]
```
This is required because of the link to the OpenAI website (for the tokenizer tool). This disclaimer (and the third-party information disclaimer) goes at the end of the last public section.

Author: There isn't a doc by that name in this repo. @gewarren is there a standard include for third-party disclamers?

---
title: "Understanding tokens"
description: "Understand how large language models (LLMs) use tokens to analyze semantic relationships and generate natural language outputs"
author: haywoodsloan
ms.topic: concept-article
ms.date: 04/05/2024

> #customer intent: As a .NET developer, I want understand how large language models (LLMs) use tokens to analyze semantic relationships and generate natural language outputs, so I may better use LLMs in my .NET projects.

Author: Sounds good!

---

# Understanding tokens

> Tokens are the words or character sets that large language models (LLMs) decompose text into. During training, an LLM analyzes the semantic relationships between these tokens, such as how commonly they're used together or if they're used in similar contexts. Once trained, an LLM uses those patterns and relationships to generate a sequence of output tokens based on the input sequence.

Author: Good point, added that the model first tokenizes the training text then analyzes the tokens.

## What are tokens?

> 

Author: I noticed most of the example concept docs did do this, so I omitted it too. Here's some of the samples I looked at:
- https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/antipatterns/strategy-antipatterns
- https://learn.microsoft.com/en-us/devops/plan/what-is-agile
- https://learn.microsoft.com/en-us/azure/azure-sql/multi-model-features?view=azuresql

> 

Author: @gewarren do you have any input here?

> 

Reviewer 2: You don't need to add that IMO.

The first step when training an LLM or inputting a prompt is tokenization. The text is broken down into a set of strings, known as _tokens_. These tokens can be distinct words, character sets, or combinations of words and punctuation. The set of unique tokens a model has been trained on is referred to as its _vocabulary_.

For example, consider the following sentence:

> I heard a dog bark loudly at a cat

This text could be tokenized as:

- I
- heard
- a
- dog
- bark
- loudly
- at
- a
- cat

Only the first occurrence of the "a" token will be added to the model's vocabulary. All other occurrences can be represented by that first token.

With a sufficiently large set of training text, a vocabulary of many thousands of tokens can be compiled.

> ## How are tokens determined?

Reviewer 3: ## Common tokenization methods

The specific tokenization method varies by LLM. Some common tokenization methods include:

- Word tokenization (text is split into individual words based on a delimiter)
- Character tokenization (text is split into individual characters)
- Subword tokenization (text is split into partial words or character sets)

For example, the GPT models, developed by OpenAI, use _Byte-Pair Encoding_ (BPE) which is a type of subword tokenization. OpenAI provides [a tool to visualize how text will be tokenized](https://platform.openai.com/tokenizer).

There are benefits and disadvantages to each tokenization method:

> **Smaller tokens (character or subword tokenization)**

Reviewer 3: Consider redoing as a table or columns.

Author: reformatted into a table, had to use some inline HTML to put a list inside the table

Reviewer 2: For bullets in a table, I just use - and `<br/>`. Anything else might get flagged at build time.

Author: Thanks for the tip. I revised the table to just have `<br>` tags

- _Pros_:
  - Allows the model to handle a wider range of inputs, such as unknown words, typos, or complex syntax
  - May allow the vocabulary size to be reduced, requiring less memory resources
- _Cons_:
  - A given text will be broken into more tokens requiring additional computational resources while processing
  - Given a fixed token limit the maximum length of the model's input and output will be smaller

**Larger tokens (word tokenization)**

- _Pros_:
  - A given text will be broken into fewer token requiring less computational resources while processing

>   - Given the same token limit the maximum size of the model's input and output will be larger

Reviewer 2: ```suggestion
  - Given the same token limit, the maximum size of the model's input and output is larger
```

- _Cons_:

>   - May cause an increased vocabulary size, requiring more memory resources

Reviewer 2: ```suggestion
  - Might cause an increased vocabulary size, requiring more memory resources
```

  - Can limit the models ability to handle unknown words, typos, or complex syntax

> ## How tokens are used

Reviewer 3: ## How LLMs use tokens

> ## How are tokens used?

Reviewer 2: ```suggestion
## How tokens are used
```

> While training, after a text has been tokenized, each unique token is assigned an ID.

Reviewer 3: "After completing tokenization, LLMs assign an ID to each unique token."

> Consider the following sentence:

Reviewer 3: It needs a period to be a sentence, so either add a period or maybe say "text" instead of "sentence".

Author: Good point, I'll use text because I don't want to have to explain what happens to the "." in this example

> For example, in the sentence:

Reviewer 2: ```suggestion
Consider the following sentence:
```

> I heard a dog bark loudly at a cat

> After tokenization, the following IDs could be assigned:

Reviewer 3: "After using a word tokenization method, a model could assign token IDs like this:"

> After tokenization the following IDs could be assigned:

Reviewer 2: ```suggestion
After tokenization, the following IDs could be assigned:
```

- I (1)
- heard (2)
- a (3)
- dog (4)
- bark (5)
- loudly (6)
- at (7)
- a (the "a" token is already assigned an ID of 3)
- cat (8)

> By assigning IDs, text can be represented as a sequence of token IDs. The previous sentence would be represented as [1, 2, 3, 4, 5, 6, 7, 3, 8]. The sentence "I heard a cat" would be represented as [1, 2, 3, 8].

Reviewer 3: "After it assigns token IDs, the model can use sequences of those IDs to represent tokenized text."

> This allows text to be represented as a sequence of token IDs. The previous sentence would be represented as [1, 2, 3, 4, 5, 6, 7, 3, 8]. The sentence "I heard a cat" would be represented as [1, 2, 3, 8].

Reviewer 2: ```suggestion
By assigning IDs, text can be represented as a sequence of token IDs. The previous sentence would be represented as [1, 2, 3, 4, 5, 6, 7, 3, 8]. The sentence "I heard a cat" would be represented as [1, 2, 3, 8].
```

> As training continues, any new tokens in the training text will be added to the vocabulary and assigned an ID. For example:

Reviewer 2: ```suggestion
As training continues, any new tokens in the training text are added to the vocabulary and assigned an ID. For example:
```

- meow (9)
- run (10)

> Using these token ID sequences, the semantic relationships between the tokens can be analyzed. These relationships are represented as multi-valued numeric vectors, known as [_embeddings_](understanding-embeddings.md). Each token will be assigned an embedding based on how commonly they are used together or in similar contexts.

Reviewer 2: ```suggestion
The semantic relationships between the tokens can be analyzed using these token ID sequences. These relationships are represented as multi-valued numeric vectors, known as [_embeddings_](understanding-embeddings.md). Each token is assigned an embedding based on how commonly it's used together with or in similar contexts to the other tokens in the embedding.
```

Reviewer 2: Not sure if my extrapolation is correct here, but felt the sentence wasn't quite complete.

> During output generation, the model predicts a vector value for the next token in the sequence. This vector value is then used to select the next token from the model's vocabulary. In practice, multiple vectors are calculated using various elements of the preceding tokens' embeddings. The model then evaluates all potential tokens from these vectors and selects the most probable one to continue the sequence.

Reviewer 2: This article implies that an embedding is associated with an entire text, rather than a token. Which is correct? https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/understand-embeddings

Author: Each token has an embedding that represents its semantic attributes. Those semantic attributes are calculated by analyzing the entire text. But each embedding only represents the semantics of the associated token. This is the reference I pulled this info from: https://learn.microsoft.com/en-us/training/modules/fundamentals-generative-ai/3-language%20models#embeddings

Author: @alexwolfmsft just wanted to double check with you that my understanding is correct

Reviewer 4: Hey @haywoodsloan - circling back to this. I'm not sure off hand if the embedding is associated with the entire text or a sum of all the individual token embeddings - I started looking into this but didn't find a confirmed answer. I will try to track down someone from the OpenAI service who might know for certain.

Author: Thanks for looking into this, I think there might be a difference between how embeddings are calculated when training a model vs when generating output. I can clarify that in this doc based on what you find.

Reviewer 5: @elbruno - are you able to answer the question on whether the embedding is associated to the entire text or just a token?

Author: As I continue to research, I think I've found an answer. 

During training each token is given an embedding, these embeddings are 1-1 with the tokens (but they are calculated by analyzing the whole text). The important distinction being that each of those embeddings only represents an individual token.

After training, embeddings can be generated for larger blocks of text. The text is tokenized and the trained embeddings for each token are used to calculate an overall embedding score for the whole text. This embeddings represents the overall semantic meaning of the entire text. This is how we store and recall text memories, we match the embeddings of the question with the embedding of the memory.

So, in some sense both cases are true depending on the context. I'll add some details about how embeddings can represent larger texts once the model is trained.

> During output generation, the model predicts a vector value for the next token in the sequence. This vector value is then used to select the next token from the model's vocabulary. In practice, multiple vectors will be calculated using various elemets of the preceding tokens' embeddings. The model will then evaluate all potential tokens from these vectors and select the most probable one to continue the sequence.

Reviewer 2: ```suggestion
During output generation, the model predicts a vector value for the next token in the sequence. This vector value is then used to select the next token from the model's vocabulary. In practice, multiple vectors are calculated using various elements of the preceding tokens' embeddings. The model then evaluates all potential tokens from these vectors and selects the most probable one to continue the sequence.
```

Output generation is an iterative operation. The predicted token is appended to the sequence so far and used as the input for the next iteration, building the final output one token at a time.

### Token limits

> LLM's will have limits on the maximum number of tokens that can be used as input context or generated as output. However, the limits for input context and output generation may differ.

Reviewer 2: ```suggestion
LLMs have limits on the maximum number of tokens that can be used as input context or generated as output. However, the limits for input context and output generation can differ.
```

> For example, GPT-4 supports up to 128,000 tokens of input context but only 4,096 tokens of output.

Reviewer 3: I've seen other token limits stated for GPT-4 (8,192 and 32,676). If the limits here are correct, I'll need to update a sentence in one of my docs (and we might check if there's already an open GitHub issue to conform the statements of token limits across docs).

Author: Looks like I grabbed the limit for the turbo model by mistake. It should be 8,192. Also, after doing some more digging, I found that GPT combines the input and output tokens into a single context token limit. I'll update this section to reflect that.

> Together a model's token limit and tokenization method determine the maximum length of text that can be provided as input or generated as output.

Reviewer 3: This is a nice insight - perhaps use the running example to provide a contrasting pair of examples (such as character-based and word-based)?

Author: Added an example of how many input and output tokens would be used/available with word or character tokenzation.

> Together a model's token limit and tokenization method will determine the maximum length of text that can be provided as input or generated as output.

Reviewer 2: ```suggestion
Together a model's token limit and tokenization method determine the maximum length of text that can be provided as input or generated as output.
```

### Token based pricing and rate limiting

> Generative AI services will often use token based pricing. The cost of each request will depend on the number of input and output tokens. The pricing may differ between input and output. For example, see [Azure OpenAI Service pricing](https://azure.microsoft.com/en-us/pricing/details/cognitive-services/openai-service/).

Reviewer 2: ```suggestion
Generative AI services often use token-based pricing. The cost of each request depends on the number of input and output tokens. The pricing might differ between input and output. For example, see [Azure OpenAI Service pricing](https://azure.microsoft.com/pricing/details/cognitive-services/openai-service/).
```

> Generative AI services might also have limits on the maximum number of tokens per minute (TPM). These rate limits can vary depending on the service region and LLM. For example, see [Azure OpenAI Service quotas and limits](/azure/ai-services/openai/quotas-limits#regional-quota-limits).

Reviewer 3: "For more information about specific regions, see [Azure OpenAI Service quotas and limits](/azure/ai-services/openai/quotas-limits#regional-quota-limits)."

> Generative AI services may also have limits on the maximum number of Tokens-Per-Minute (TPM). These rate limits can vary depending on the service region and LLM. For example, see [Azure OpenAI Service quotas and limits](https://learn.microsoft.com/en-us/azure/ai-services/openai/quotas-limits#regional-quota-limits).

Reviewer 2: ```suggestion
Generative AI services might also have limits on the maximum number of tokens per minute (TPM). These rate limits can vary depending on the service region and LLM. For example, see [Azure OpenAI Service quotas and limits](/azure/ai-services/openai/quotas-limits#regional-quota-limits).
```

## Related content

- [How Generative AI and LLMs work](how-genai-and-llms-work.md)
- [Understanding embeddings](understanding-embeddings.md)

> - [Working with vector databases](working-with-vector-dbs.md)

Reviewer 3: Is there a trailing line? Add one if not.

Author: yep!
